Complete the README and begin CONTRIBUTING (#16)
meilisearch/cloud-providers: CONTRIBUTING.md

# Contributing <!-- omit in TOC -->

> First of all, thank you for contributing to Meilisearch! This document aims to provide everything you need to know to contribute to Meilisearch and its different integrations.

Reviewer: ```suggestion
First of all, thank you for contributing to Meilisearch! This document aims to provide everything you need to know to contribute to this Meilisearch integration.
```

- [Assumptions](#assumptions)
- [How to Contribute](#how-to-contribute)
- [Development Workflow](#development-workflow)
- [Git Guidelines](#git-guidelines)
- [Release Process (for the internal team only)](#release-process-for-internal-team-only)

## Assumptions

> 1. **You're familiar with [GitHub](https://github.com) and the [Pull Request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-requests)(PR) workflow.**

Reviewer: ```suggestion
1. **You're familiar with [GitHub](https://github.com) and the [Pull Request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests)(PR) workflow.**
```

3. [Create a new Git branch](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-and-deleting-branches-within-your-repository).
4. Make the changes on your branch.
5. [Submit the branch as a PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request-from-a-fork) pointing to the `main` branch of the main cloud-providers repository. A maintainer should comment and/or review your Pull Request within a few days. Although depending on the circumstances, it may take longer.<br>
 We do not enforce a naming convention for the PRs, but **please use something descriptive of your changes**, keeping in mind that the title of your PR will be automatically added to the following [release changelog](https://github.com/meilisearch/cloud-providers/releases/).

You can check out the more extended, complete guideline documentation [here](https://github.com/meilisearch/.github/blob/main/Hacktoberfest_2022_contributors_guidelines.md).

## Development Workflow

### Setup <!-- omit in toc -->

1. Install the [Packer CLI](https://developer.hashicorp.com/packer/downloads)
2. Set up all your needed credentials

> - Aws:

Reviewer: ```suggestion
- AWS:
```

Obtain your [AWS access keys](https://docs.aws.amazon.com/keyspaces/latest/devguide/access.credentials.html) and set it in your environment:
  ``` bash
    export AWS_ACCESS_KEY_ID="YOUR_ACCESS_KEY"
    export AWS_SECRET_ACCESS_KEY="YOUR_SECRET_KEY"
  ```
- DigitalOcean:
Before running any script, make sure to [obtain a DigitalOcean API Token](https://www.digitalocean.com/docs/apis-clis/api/create-personal-access-token/) and set it in your environment:
  ```bash
  export DIGITALOCEAN_ACCESS_TOKEN="XxXxxxxXXxxXXxxXXxxxXXXxXxXxXX"
  ```

> - Gcp

Reviewer: ```suggestion
- GCP:
```

``` bash
packer fmt .
```

Use this command to validate the syntax and configuration of the template:
``` bash
packer validate .
```

### Build the images <!-- omit in toc -->

⚠ Please be aware that this command will create all new Meilisearch images on all platforms.

This command will execute the template and build all images in parallel:
``` bash
packer build meilisearch.pkr.hcl
```

You can build one or more images by specifying their name, take the first parameter given to `source` block and add it the following wildcard regex: `.*`.
Example:
``` bash
packer build -only 'amazon-ebs.*' .
```

### Add a new provider

There are multiple providers available for Packer, you can check the [External Plugins Documentation](https://developer.hashicorp.com/packer/plugins)

To add a provider correctly to the list of existing providers, proceed as follows:
1. Add your provider's code to the main `meilisearch.pkr.hcl` file as expected.
2. Add the provider name to the [README](README.md#-providers-available).
3. Add a manual trigger for your new provider to the CI by adding it to the list of suppliers in the `option` block of [manual_publish.yml](/manual_publish.yml).

> 

Reviewer: I think the first one should be:

1. Explain through detailed instructions using text/images/video the steps you've used to create and configure the provider to be able to receive new images.

- The branch related to the PR must be **up-to-date with `main`** before merging. If it's not, you have to rebase your branch. Check out this [quick tutorial](https://gist.github.com/curquiza/5f7ce615f85331f083cd467fc4e19398) to successfully apply the rebase from a forked repository.
- All PRs must be reviewed and approved by at least one maintainer.
- The PR title should be accurate and descriptive of the changes.

## Release Process (for the internal team only)

Meilisearch tools follow the [Semantic Versioning Convention](https://semver.org/).

The release tags of this package follow exactly the Meilisearch versions.<br>
For example, the `v1.1.1` tag in this repository corresponds to the image running Meilisearch `v1.1.1`.

**Please follow the steps in the next sections carefully before any release.**

### Automated Publishing <!-- omit in TOC -->

Building and deploying all new Meilisearch images on all platforms is done automatically by a github action when a new tag is pushed to main.

### Release, Build and Publish <!-- omit in TOC -->

⚠️ The image should never be published with a `RC` version of Meilisearch.

> 1. In the file [`meilisearch.pkr.hcl`](meilisearch.pkr.hcl), update the value of the `default` variable in the block `meilisearch_version` with the new version of Meilisearch you want to release, in the format: `vX.X.X`.
> 
> 2. Commit your changes to a new branch.
> 
> 3. Open a PR from the branch where changes were done and merge it.
> 
> 4. Create a git tag on the last `main` commit:

Reviewer: ```suggestion
1. In the file [`meilisearch.pkr.hcl`](meilisearch.pkr.hcl), update the value of the `default` variable in the block `meilisearch_version` with the new version of Meilisearch you want to release, in the format: `vX.X.X`.
2. Commit your changes to a new branch.
3. Open a PR from the branch where changes were done and merge it.
4. Create a git tag on the last `main` commit:
```

```bash
git checkout main
git pull origin main
git tag vX.X.X
git push origin vX.X.X
```
This will start building and publishing all images on all platforms.

### Publish each image manually <!-- omit in TOC -->

In case a publication did not go well, you can manually launch each building image separately in the GitHub interface.
- Click on the Action tab.
- Choose the "Publish Manually" action on the left sidebar.
- On the right side of the screen, select the `Use workflow from` panel.
- On `Use workflow from` keep `Branch: main`.
- On `Choose the provider you want to publish` select the provider you want to re-deploy.

> ### Publish the DO Image <!-- omit in TOC -->

Reviewer: ```suggestion
### Publish the DigitalOcean image <!-- omit in TOC -->
```
